File: AI/aithread.cpp
#include "aithread.h"
#include "ai.h"
#include "logic.h"
#include <iostream>

AIThread::AIThread(const size_t& instanceId, AI& ai) : m_instanceId(instanceId), 
m_ai(ai), m_master(ai.getMaster()), m_hasStarted(false), m_thread(runThread,this),  
m_rootNbSimu(0), m_hasEvents(false), m_lock(m_mutex), m_mustStop(false)
{
}

void AIThread::start()
{
    m_mustStop = false;
    _pushEvent(0, Event::START);
}

void AIThread::stop()
{
    m_mustStop = true;
}

bool AIThread::hasStarted() const
{
    return m_hasStarted;
}

void AIThread::pushAction(const size_t& action)
{
    m_crtActions.push_back(action);
    m_ai.pushCrtAction(action, m_instanceId);
}

bool AIThread::mustStop()
{
    return m_mustStop.load();
}

void AIThread::computeActions()
{
    N_DarkLogic::DarkLogic::getActions(m_instanceId);
}

void AIThread::updateLogic(const size_t& actionId)
{
    N_DarkLogic::DarkLogic::apply(m_instanceId,actionId);
    m_crtActions.clear();
}

unsigned char AIThread::instanceId() const
{
    return m_instanceId;
}

void AIThread::setRootNbSimu(const size_t& nbSimu)
{
    m_rootNbSimu = nbSimu;
}

size_t AIThread::getRootNbSimu() const
{
    return m_rootNbSimu;
}

void AIThread::incrRootNbSimu()
{
    m_rootNbSimu++;
}

void AIThread::_start()
{
    //std::cout << "[DEBUG] Start thread:"<<m_instanceId << std::endl;
    m_hasStarted = true;
    N_DarkLogic::DarkLogic::getActions(m_instanceId);

    //compute value of given nodes
    while (!mustStop())
    {
        m_ai.explore(m_crtActions);
    }

    //inform master that this thread has finished
    _stop();
}

void AIThread::_stop()
{
    m_hasStarted = false;
    //std::cout << "[DEBUG] Stop thread:" << m_instanceId << std::endl;
    m_master->stopFromThread(m_instanceId);
}

[[noreturn]] void AIThread::_run()
{
    while (true)
    {
        while (!m_hasEvents)
        {
            m_condition_var.wait(m_lock);
        }
        if(m_eventQueue.size())
        {
            //Consuming event
            switch (m_eventQueue.front().type())
            {
            case Event::START:
            {
                _start();
                break;
            }
            case Event::STOP:
            {
                _stop();
                break;
            }
            case Event::STOP_THREAD:
            {
                break;
            }
            }
            m_eventQueue.pop();
        }
        m_hasEvents = false;
    }
}

void AIThread::_pushEvent(const unsigned char threadIdx_, Event::EventEnum type_)
{
    m_mutex.lock();
    m_eventQueue.push(Event(threadIdx_, type_));
    m_mutex.unlock();
    m_hasEvents = true;
    m_condition_var.notify_all();
}

void runThread(AIThread* thread)
{
    thread->_run();
}

File: PyDarkLogic/AI/aithread.py
import os

path = os.getcwd()
import sys

sys.path.append(path + "\..\..\Lib")
from DarkLogic import DarkLogic
from threading import Thread, Condition, Lock
from AI.event import Event
from AI.dbnode import DbNode


# friend method
def _start(aiThread):
    aiThread.start_()


# friend method
def _stop(aiThread):
    aiThread.stop_()


class AIThread(Thread):
    _switcher = {Event.EventEnum.START: _start,
                 Event.EventEnum.STOP: _stop}

    def __init__(self, instanceId, ai, master):
        Thread.__init__(self)

        self._instanceId = instanceId
        self._ai = ai
        self._master = master
        self._crtActions = DbNode()

        # start/stop thread
        self._hasStarted = False
        self._mustStop = False
        self._mutexStop = Lock()

        # handling eventss
        self._hasEvents = False
        self._mutex = Lock()
        self._condition_var = Condition(self._mutex)
        self._eventQueue = []

    def hasStarted(self):
        return self._hasStarted

    def pushAction(self, parent, action):
        idx = self._crtActions.push(parent, action)
        valSon = self._ai.realValueOfActions([parent, action])
        self._crtActions.updateValue(idx, valSon)

    def mustStop(self, action=None):
        self._mutexStop.acquire()
        ret = self._mustStop
        self._mutexStop.release()
        return ret

    def computeActions(self):
        DarkLogic.getActions(self._instanceId)

    def updateLogic(self, actionId):
        DarkLogic.apply(self._instanceId, actionId)
        self._crtActions.clear()

    def instanceId(self):
        return self._instanceId

    def start_(self):
        self._hasStarted = True
        DarkLogic.getActions(self._instanceId)

        # compute value of given nodes
        while not self.mustStop():
            self._ai.explore(self._crtActions, self._instanceId)

        # inform master that this thread has finished
        self.stop_()

    def stop_(self):
        self._hasStarted = False
        self._master.stopFromThread_(self._instanceId)

    def run(self):
        while True:
            while not self._hasEvents:
                with self._condition_var:
                    self._condition_var.wait()
            if len(self._eventQueue):
                # Consuming event
                AIThread._switcher[self._eventQueue[0].type()](self)
                self._eventQueue.pop(0)
            self._hasEvents = False
            if self.mustStop():
                self._eventQueue.clear()
                break

    def _pushEvent(self, threadIdx_, type_):
        self._mutex.acquire()
        self._eventQueue.append(Event(threadIdx_, type_))
        self._mutex.release()
        self._hasEvents = True
        with self._condition_var:
            self._condition_var.notify_all()

    def _stop(self):
        self._mutexStop.acquire()
        self._mustStop = True
        self._mutexStop.release()

    def _start(self):
        # print("start ai thread "+str(self._instanceId))
        self._mustStop = False
        if not self.is_alive():
            self.start()
        self._pushEvent(0, Event.EventEnum.START)
